torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -295,6 +296,13 @@ class ProcessGroupNCCL : public ProcessGroup {
       addCallback(std::bind(
           [&](std::function<at::IValue(void)> cb) {
             try {
+              // Do not free the input tensors before their usages on
+              // futureNCCLCallbackStream_ finish.
```

Comment:
There is only one tensor in the vector, so let's just do something like this to be explicit:

```
auto tensors = value_.toTensorVector();
TORCH_INTERNAL_ASSERT(1, tensors.size());
c10::cuda::CUDACachingAllocator::recordStream(
                    tensors[0].storage().data_ptr(), *futureNCCLCallbackStream_);
```

Reply:
Done. The size of tensors is indeed 1.

However, realized that the change here cannot pass test_accumulate_gradients_no_sync_allreduce_with_then_hook which invokes then -- the execution fails after this , with an error of "Timing out after 100 seconds and killing subprocesses".

I even tried to place this added code after "fut->markCompleted(at::IValue(cb()));" or even right before returning the future in this lambda, and tried to comment out recordStream call. Didn't work.

Reply:
Can you debug to see why the test is timing out? You can probably add some log statements to the unit test to see where it is getting stuck. Another option is to attach gdb to the child processes when the test is stuck.

Reply:
I added LOG statements to debug before asking actually, but all the logs in this function could be printed out, so it's not timed out in this function. Will try gdb.

Reply:
It turns out that the type of future value here is pyobject rather than TensorList, but a type casting will require pybind and introduce Python dependency, which should not be allowed in c10d library.

I have considered creating a util function in a separate file to support this type casting, and then placing it under torch/csrc directory where python dependency is allowed. However, torch/csrc has a dependency on c10d, so this will create a circular dependency.

Per offline discussion, added a `record_stream_cb_` member to FutureNCCL, and the default value is nullptr. A default `record_stream_cb_` implementation is added to `get_future` binding in init.cpp, where Python dependency is allowed. record_stream_cb_ can be customized by caller via `setRecordStreamCallback`, so here the future value does not have to be a tensor vector.